# Estimador Máximo Verosímil

_Miércoles 22 de abril_


Ejemplo 1. Lanzamos una moneda **100** veces, sabiendo que se obtienen **55** soles. Encontrar el EMV para la probabilidad $p$ de obtener un sol en un sólo lanzamiento.

Sea
$$ X: \text{el número de soles que obtienen al lanzar una moneda} $$
tenemos que $X\sim \text{Bin}(100,p)$, entonces
$$f\_{X}(x;p) = \binom{100}{55}p^{55}(1-p)^{100-55}$$


In [1]:
from scipy.stats import binom
import numpy as np

# Vamos a utilizar la libreria SymPy para matemáticas símbolica
import sympy

from sympy.abc import x # importa la variable simbólica x

https://www.sympy.org/en/features.html


In [2]:
# Vamos a definir a p como una variable simbólica positiva (de manera manual)
p = sympy.symbols('p', positiva=True)
# Definomps la función de verosimilitud 
# omitimos la constante binom{100}{55}
f = p**55 * (1-p)**(100-55)

# Ahora derivamos f con respecto a p
df = sympy.diff(f,p)

# Para encontrar el estimador, igualamos la derivada a cero y despejar p
phat = sympy.solve(df,p)[1] #df=0 para encontra p y [1] para la segunda solución ya que [0] es cero

print("El estimador de máxima verosimilitud para p es", phat)

El estimador de máxima verosimilitud para p es 11/20


Entonces el valor coincide con la proporción de soles.


### Ejemplo 2.


In [3]:
# Simualaremos una muestra de tamaño 100
# n=1 : un ensayo de Bernoullo
# p: proba de éxito
# size: genera 100 observaciones
r = binom.rvs(n=1,p=0.7,size=100) # Genera v.a. Binomiales
r

array([0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1])

In [4]:
# Vamos a calcular EMV

# Definimos la función de verosimilitud para una sola observación
f = p**x * (1-p)**(1-x) # x actua como variable simbólica
# Calculamos el producto de todos los elementos
# subs : para sustituir la variable simbolica por valor
J = np.prod([f.subs(x,i) for i in r]) 

# calculamos el log de J 
# sympy.log(J) # log(J)
logJ = sympy.expand_log(sympy.log(J))

phat = sympy.solve(sympy.diff(logJ,p),p)[0]
print("El estimador de máxima verosimilitud para p es", phat)

El estimador de máxima verosimilitud para p es 7/10


In [5]:
# Podemos revisar las soluciones con:
sol = sympy.solve(sympy.diff(logJ,p),p)
sol

[7/10]

### Ejemplo

El tiempo de retardo para pacientes con enfermedad coronaria potencial se define como la duración entre el comienzo de un síntoma agudo cardiovascular y la llegada a la unidad de emergencia.

Se supone que el tiempo de retardo sigue una distribución exponencial con parámetro $\theta$.

Se tienen los siguientes registros de retardo (en minutos) en la clinica para los primeros 20 pacientes:

$$ 525, \ 719, \ 2880, \ 150, \ 30, \ 251, \ 45, \ 858, \ 15, \ 47, \ 90, \ 56, \ 68, \ 6, \ 189, \ 180, \ 60, \ 60, \ 294, \ 747 $$

- Encontrar el estimador por el método de momentos para la media
- (Python) Encontrar el valor del estimador con los datos dados.

Se tiene que
$$X : \text{tiempo de retardo (en minutos)} $$
y sigue $\text{Exp}(\theta)$.

Recordemos que $\mathbb{E}[X]=\mu=1/\theta$ y $\text{Var}(X)=1/\theta^{2}$

Por el metodo de momentos, se tiene

$$ \mathbb{E}[X] = \bar{X}$$
En este caso el estiamador para la media es la media muestral.


In [6]:
# Datos observados
datos = np.array([525, 719, 2880, 150, 30, 251, 45, 858, 15, 47, 90, 56, 68, 6, 189, 180, 60, 60, 294, 747])
# Tamaño de la muestra
n = len(datos)

# El estimador es:
media_muestral = np.mean(datos)
print("La media muestral es:", media_muestral)

La media muestral es: 363.5


Se tiene que $\theta=\frac{1}{\mathbb{E}[X]}$, entonces

$$ \hat{\theta} = \frac{1}{\bar{X}} \approx 0.00275$$
por minuto.

Por lo tanto, estimamos que el **tiempo promedio de retardo** entre el inicio de sintomas cardiovasculares y la llegada a la unidad de emergencia es de aproximadamente 363.5 minutos.


**Ejercicio (Pyton)**. Suponga que la vida util de los focos de cierta marca, se modela mediante una distribución expoencial de parametro $\theta$ (desconocido). Probamos 5 focos y encontramos que tienen una vida útil de $2,3,1,3$ y $5$ años, respectivamente. ¿Cuál es el MLE para $\theta$?

Sabemos que

$$ f(x*i;\theta) = \theta \exp(-\theta x*{i}) , x\ge 0, \theta>0 $$

$$ L(x,\theta) = \prod*{i=1}^{n} f(x_i;\theta) = \prod*{i=1}^{n} \theta \exp(-\theta x\_{i}) = \theta^{n}e^{-\theta \sum x_i}$$

$$ \ln L(x,\theta) = n\ln(\theta)-\theta\sum\_{i}x_i$$

$$\frac{d \ln L(x,\theta)}{d\theta} = \frac{n}{\theta}-\sum_{i}x_i = 0$$

$$ \hat{\theta} = \frac{n}{\sum\_{i}x_i} = 1 / \bar{X} $$


In [8]:
# Datos: 
datos = np.array([2,3,1,3,5])
n=len(datos)

suma = np.sum(datos)
media = np.mean(datos)
media

theta = n / suma #(1/media)

print("La media es", media)
print("El estimador de máxima verosimilitud para θ es", theta)


La media es 2.8
El estimador de máxima verosimilitud para θ es 0.35714285714285715


La tasa estimada de falla es de $0.3571$ fallas por año y se tiene que la vida media estimada de los focos es de 2.8 años


**Ejercicio.** Calcular el estimador con Sympy.


In [ ]:
datos = np.array([2,3,1,3,5])
n=len(datos)

theta = sympy.symbols('theta', positive=True)

LogL = n * sympy.log(theta) - theta * np.sum(datos)

dLogL = sympy.diff(LogL,theta)

estimador = sympy.solve(dLogL,theta)[0]

estimador_decimal = estimador.evalf()

print(estimador)
print(f"El MLE para theta es: {estimador_decimal}")


5/14
El MLE para theta es: 0.357142857142857
